### ISSUE TITLE
PVCProtectionController can be slow for larger clusters

### ISSUE THREAD

**Issue by VirrageS** [kubernetes/kubernetes#109282]
## References
* https://github.com/kubernetes/kubernetes/issues/76703
* https://github.com/kubernetes/kubernetes/issues/75980
* https://github.com/kubernetes/kubernetes/pull/80492
* https://github.com/kubernetes/design-proposals-archive/blob/main/storage/postpone-pvc-deletion-if-used-in-a-pod.md

## Overview

As described [here](https://github.com/kubernetes/design-proposals-archive/blob/main/storage/postpone-pvc-deletion-if-used-in-a-pod.md) PVC protection controller uses logic that postpones deletion of PVC in case it is used by Pod. When the Pod is deleted and PVC is marked for deletion as well (by setting `deletionTimestamp`) the PVC protection controllers ensures that all references to the PVC has been deleted and no Pod is using it.

PVC protection controller processes PVC one by one checking if there is still a Pod that has any reference to it. If not, the finalizer is removed and PVC is deleted from the cluster.

## The problem

Some time ago the [issue](https://github.com/kubernetes/kubernetes/issues/76703) was filled that described race condition in PVC protection controller. Basically the logic used and relied only on Pod's informer cache which could be potentially outdated and PVC protection controller could deleted PVC that it's still being used (has been recently attached). There was also [MR](https://github.com/kubernetes/kubernetes/pull/80492) that fixed it.

The problem with the new logic is that it has really slow/inefficient performance which is in order of `O(m * n)` (where `m` is the number of PVCs to be processed and `n` is the number of Pods in the cluster). [Someone even pointed it out](https://github.com/kubernetes/kubernetes/pull/80492#discussion_r306743513) in the MR but it wasn't sufficiently measured/analyzed. For active larger cluster 1k+ where there is a lot of Pods/PVCs being created and deleted the PVC protection controller is too slow to process everything. This causes large number of PVCs still being in the cluster what can slow down listing and scheduling but also puts unnecessary pressure on the API server and etcd.

What happens under the hood is that when the Pod is quickly removed from the cluster and PVC is marked for deletion we always will go past the informer check and make another API LIST call: https://github.com/kubernetes/kubernetes/blob/808c8f42d55c520e45eab93200768607cefc1451/pkg/controller/volume/pvcprotection/pvc_protection_controller.go#L212-L226

This means that for each PVC we will do full-blown LIST call that returns all Pods in given namespace. It's easy to foresee that this logic will not be fast enough to handle active namespaces with 1k+ Pods/PVCs.

Here is the PVC protection controller's work queue and the PVC count over 5 hours of active creation/deletion that tests this hypothesis. What happens here is that PVC protection controller is slow to delete the finalizer so the work queue and number of PVCs in the cluster only increases.
![slow](https://user-images.githubusercontent.com/3855799/161510944-cf39c3ef-3b97-4b90-94db-765c725e8f54.jpg)

This can easily blow up the cluster if no solution is implemented as a remedy.

## The (temporary) solutions

There are couple of ways to go around this problem:
1. Reduce the size of the namespace. When we list and process Pods/PVCs we operate on single namespace. If we reduce the size of the namespace the PVC protection controller will have an easier work.
2. Implement component that will automatically delete protection finalizer for PVCs which for sure will not be used. Basically if you know that the PVCs you are creating will not be reused you can simply automatically delete the finalizer without relying on the PVC protection controller.
3. Turn off protection finalizer (see https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#storageobjectinuseprotection).

Probably there is more (I'm happy to hear if you have suggestions).

But all of proposed solution don't address the original problem and only make it less probable.

## The (considered) solutions 

I've been thinking on addressing the logic problem but I couldn't come up with any good idea:
* I initially considered making PVC protection controller multi-threaded. But this doesn't reduce the overall performance issues and in some cases it could even make things worse. It would probably help in most of the cases but the underlying problem would still remain.
* Then I considered batching. Currently we process one PVC at a time. We could lower number of Pod LIST calls by processing N number of PVCs and executing single Pods LIST call. So instead of having "1 PVC -> 1 Pod LIST call" we could have eg. "20 PVC -> 1 Pod LIST call". This seems like a interesting approach but it would require changing the whole logic of the controller.

**Comment by thehackercat** [comment]
I wonder if we can just remove the logic of `AskApiserver` ？

**Comment by FillZpp** [comment]
> I wonder if we can just remove the logic of AskApiserver ？

+1, list all Pods in a namespace from apiserver is a bad idea, it will be terrible when there are a large number of Pods and PVCs.

If we are concerning that some Pods using this PVC may have not synced into cache yet, how about multiple check from informer with a little backoff:

1. For a terminating PVC that find not in-use from `askInformer`, record into a memory cache, and then `queue.AddAfter` with a little duration (could be 100ms~1s)
2. If the PVC is still not in-use from `askInformer` in the next reconcile, then remove its finalizer. Also we can do this more times if ppl think it is not enough to check twice from informer.

WDYT @msau42 @jsafrane @matte21 

If this is an acceptable change, I would be happy to raise a Draft PR for it.

**Comment by mborsz** [comment]
Naive solution for this problem would be to change isBeingUsed to:
* use live apiserver to get the up to date resourceVersion for pods resource (e.g. by doing a LIST call with limit=1 that will return both random element (to be ignored) and up to date resource version for pods resource)
* waiting until the pod informer reaches a given resourceVersion (probably just comparing pvcInformer.Informer().LastSyncResourceVersion() with the rv received above should be enough).
* using informer without live apiserver (we have a guarantee that the informer is up to date).

This way we can remove risk that we will make operation based on the stale informer state and be able to perform check quickly as we can avoid expensive list, use podIndexer's index instead of O(n) lookup we do in askAPIServer.

@jsafrane @VirrageS, WDYT?

**Comment by VirrageS** [comment]
I might be missing something about this proposal. But assuming we do "apiserver [call] to get the up to date resourceVersion for pods resource" and "waiting until the pod informer reaches a given resourceVersion" during this waiting the resourceVersion could be already updated and we will be waiting for something that is already old and outdated.

**Comment by mborsz** [comment]
Probably I don't understand the concern. I was thinking about waiting for **at least** given resource version (probably I should be more clear about that). If after listing we will find that the informer is already in greater resource version that we requested that's good -- we can stop waiting immediately. If we never see specific resourceVersion, but we see larger resourceVersions, we are good as well, as the resource version for specific resource type only grows.

**Comment by VirrageS** [comment]
Hm, let me understand the proposal. So whenever we are about to delete the PVC we do:
1. "a LIST call with limit=1 that will return both random element",
2. check `resourceVersion` from previous step,
3. we wait for the informer to be up to date (it must reach at least `resourceVersion` from previous step),
4. once the informer reaches the expected state we use it to list and check all the pods.

Is that the idea?

For me it looks really reasonable. I don't see any flaws with that.

**Comment by mborsz** [comment]
Yes, this is the idea.

Potentially in "4. once the informer reaches the expected state we use it to list and check all the pods." we can simply use index instead of iterating all the pods to make it more performant.

If the concern is that "what happens if the after point "3." someone adds the pod using PVC, but this is somewhat consistent with LIST pods approach -- we check state at some point and if it changes after the list call starts we won't see it. I presume that there is some mechanism that prevents PVCs waiting for deletion to be bound to the pods that prevents creating new pods after PVC is deleted, so this should be OK.

**Comment by mborsz** [comment]
In fact the solution proposed by me in https://github.com/kubernetes/kubernetes/issues/109282#issuecomment-1263207774 isn't correct as the resourceVersion in LIST call and in watch can differ much if there are no changes to a given resource type, e.g.:
* WATCH on pods observed all pod changes up to RV=50
* LIST returns RV=100, as there are some changes to e.g. replicasets or configmaps using the same etcd
* As long as there is no pod change between RV=50 and RV=100 or we don't get a watch bookmark (every ~1m?), the watch will be still on RV=50, so waiting for it to be on RV=100 (as described in algorithm) may take significant time.

**Comment by mattcary** [comment]
Do [watch bookmarks](https://kubernetes.io/docs/reference/using-api/api-concepts/#watch-bookmarks) help with the resource version problem mentioned above? The semantics of bookmarks seem a little vague, but maybe we can depend on a bookmark being sent frequently enough to tell our watch is up-to-date.

**Comment by VirrageS** [comment]
From the docs:
> As a client, you can request BOOKMARK events by setting the allowWatchBookmarks=true query parameter to a watch request, but you shouldn't assume bookmarks are returned at any specific interval, nor can clients assume that the API server will send any BOOKMARK event even when requested.

So I don't think this is reliable source, and here we are dealing with something that MUST work all the times.

**Comment by mattcary** [comment]
My understanding is that when a bookmark event comes, the resource version there will tell us for sure if we are up-to-date on the pod list.

This is the same as @mborsz's proposal, except that if the bookmark events come more frequently than the watch RV gets updated, then it fixes the "significant time" problem @mboarsz mentioned.

Even if there isn't a strict frequency for the bookmark events to come, if they're frequent enough in practice that would probably be fine?

**Comment by seamongit** [comment]
hi~any update?

**Comment by msau42** [comment]
One possible mitigation is to use `Node.Status.VolumesInUse` as a signal instead of `Pods`. However we would need to think through possible race conditions (ie pods getting added to nodes)

**Comment by smileusd** [comment]
I also consider to recored the status from nodes, but i not sure how about multiple pod use the same pvc @msau42

**Comment by smileusd** [comment]
And i think Node.Status.VolumesInUse is complicated and bind too much condition, a new field may be more clear?

**Comment by leiyiz** [comment]
> As long as there is no pod change

would it solves this issue if we do a pod update immediately after getting the result from the `List()` to guarantee there's pod update happening

**Comment by leiyiz** [comment]
can we do a timestamp based caching on the live list results? basically:
- when we do live lists, we grab the current timestamp from the api server with the [time api](https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/time.go) right before we call the live list, then cache the results for the following delete calls for a period (cache TTL, say 1 sec).
- in the following deletevolumes, we use the cache if the TTL hasn't passed and if the volume's deletionTimestamp is a previous time compared to the time when the list cache was generated. The cache will be refreshed if either of those conditions are not true.

I am not sure if there's any guarantee within the api server of the timestamp ordering but if so I haven't figured out a scenario that would cause a race condition.

**Comment by mattcary** [comment]
We recently saw a similar issue. It was associated with many pods being deleted at once. That suggests that batching requests so that we amortize the cost of the loop over all the deleted pods would work -- I'm not sure if that's easy though in the current architecture of the controller.

**Comment by sunnylovestiramisu** [comment]
Discussed offline: one possible mitigation is to give up on correctness and make this best effort by listing resources from informer.

**Comment by mattcary** [comment]
> Discussed offline: one possible mitigation is to give up on correctness and make this best effort by listing resources from informer. But we need to evaluate the behavior of what happens under that condition.

Wouldn't that just be going around in circles? https://github.com/kubernetes/kubernetes/pull/80492

The problem is listing every pod for every PVC. There is a lot of repeated information.

I think we know that once a PVC has a deletion timestamp, no new pods will attach to it? So if we've listed all pods & used PVCs at time t, then that list will be valid for all PVCs whose deletion timestamp is < t, and it's not necessary to hit the API server. 

This doesn't change much when PVCs are deleted infrequently. But the scaling issues we saw is when 10s of thousands of PVCs and their pods are deleted at once.

**Comment by msau42** [comment]
When kubelet admits a pod, it does a live Get() of PVC so it will error if PVC has a deletion timestamp on it. So new Pods won't start.

Using a timestamp based caching approach (similar to what @leiyiz proposed) of Pods and comparing to PVC deletion timestamp sounds plausible. Although in the case of thousands of PVCs deleted at once, I'm not sure how effective it will be.

**Comment by mattcary** [comment]
Several of us in GKE have been discussing this and have come up with the following.

There’s consensus that the chance of the informer being stale is real, so any solution needs a live pod listing at some point -- a pathological case pointed out to us is where the apiserver becomes partitioned from etcd, so no informer updates come through, because there is no etcd watch to supply them.  Then, a controller reconciles on the stale data, and etcd becomes available just in time for the write that actuates the reconciliation to be handled.

There’s two ideas we have: batching work, and adding a pod list cache. We think the batching idea is better so I’ll start with that. Any comments/reactions or should we go ahead and do this?

**Batching**

Algorithm sketch
* In [processNextWorkItem](https://github.com/kubernetes/kubernetes/blob/release-1.27/pkg/controller/volume/pvcprotection/pvc_protection_controller.go#L122), pull all available items from the queue.
* call processPVC, adding a lazyLivePodList to the call
* a lazyLivePodList (naming TBD) is a type that returns a list of pods. The first time it calls the api server. Subsequent calls return the first results
    - This is needed because most times processPVC is called, no live pod list is done, because PVCs are in use and we just work off the informer

Open questions
* The controller uses a rate limiting workqueue. Do we have to change that to batch enough PVCs together? Probably we should write a PoC first and see how it works. The main point of rate limiting seems to be to not busy-retry failures, so as long as we have a rate limiter in the process work item frequency it should be fine.

**Cache**

This method assumes that deletion timestamps are increasing. We assume we can’t have a pvc A with deletion timestamp X in our informer, then do a pod listing, then have a pod start using pvc B, and then have B get a deletion timestamp X-1.

Algorithm sketch
* Using the existing PVC informer, watch for PVCs with deletion timestamps and cache the most recent timestamp in PVC protection controller memory.
* When a live pod listing is performed, mark it with the most recent PVC deletion timestamp (as seen by PVC informer).
* When [checking if a PVC marked for deletion is in use](https://github.com/kubernetes/kubernetes/blob/release-1.27/pkg/controller/volume/pvcprotection/pvc_protection_controller.go#L213), 
  - check if it is in use according to the pod informer cache; if so return in use
  - check if the deletion timestamp is less than or equal to the timestamp of the cached live pod listing; if so return in use
  - only if the previous checks fails, do a live pod listing and return according to that.

With this method, a live pod listing is only done for a newly deleted PVC. If there is a backlog of deleted PVCs, once one live listing is done, no new listing will be done until the backlog is cleaned out, which will happen quickly since the pod list is in-memory.

**Comment by mattcary** [comment]
After a brief discussion with @liggitt on the [sig-api-machinery slack](https://kubernetes.slack.com/archives/C0EG7JC6T/p1711418407770279?thread_ts=1711053796.643179&cid=C0EG7JC6T), the batching idea seems good.

The nuances are (1) to ensure we keep the batch groups per-namespace so we keep the same filtering on pods, and (2) rate limiting should happen on the queue that makes api server calls and requeues errors. So that means we won't rate-limit the pvc informer queue. On pvc notification we add it to a per-namespace map, add the namespace to a workqueue that is rate-limited.

Please respond on this thread or reach out on slack if you have any questions or comments!

**Comment by jsafrane** [comment]
Do you plan to cache every namespace that has PVCs or do you want to release caches for old namespaces (and how old)?

**Comment by mattcary** [comment]
With the batching method there is no cache to track --- that's its primary advantage.

Instead of caching, we make just one live api server call per batch, so it gets the same scaling benefits as using a cache.

**Comment by liggitt** [comment]
> With the batching method there is no cache to track --- that's its primary advantage.
> 
> Instead of caching, we make just one live api server call per batch, so it gets the same scaling benefits as using a cache.

Um... every informer maintains a cache internally... if the controller is already maintaining a cache of pods (which kube-controller-manager is), a first pass checking use in cached pods would be *fast* and require no API server lookups. Only if the cached pods indicated no use would a live lookup be required.

**Comment by mattcary** [comment]
It's probably buried in the comments above, but the current implementation does look at the informer cache and only if necessary do a live api check. We wouldn't be changing that behavior, ie if all pvcs are in use according to the informer cache then no live api call will be performed.

The comment about the cache is around the second option in [this comment](https://github.com/kubernetes/kubernetes/issues/109282#issuecomment-2013443881) which it sounds like we've decided against.

**Comment by liggitt** [comment]
> We wouldn't be changing that behavior, ie if all pvcs are in use according to the informer cache then no live api call will be performed.

Sounds good, just double-checking

**Comment by pwschuurman** [comment]
One comment about the **Cache** approach.

> This method assumes that deletion timestamps are increasing. We assume we can’t have a pvc A with deletion timestamp X in our informer, then do a pod listing, then have a pod start using pvc B, and then have B get a deletion timestamp X-1.

Is there risk that clock skew between APIServers can break this assumption?

Consider the sequence of events:

1. APIServer1 leader is running at relative time X.
2. PVC A is deleted and assigned deletionTimestamp X+1.
3. PVCProtectionController admits PVC A to its workqueue and caches X+1 as the most recent timestamp.
4. PVCProtectionController performs a list Pods call, marking this cached result with X+1.
5. APIServer1 leader election fails over to APIServer2. APIServer2 has relative time Y. Y has a clock skew S, where Y=X-S
6. PVC B is deleted and assigned deletionTimestamp Y+2. PVC B resource has a deletionTimestamp of X+2-S
7. PVCProtectionController processes its workqueue. It sees that PVC A can be deleted (correctly). Depending on S, if S > 1, PVC B appears to have been deleted before X+1, despite having been deleted after X+1 (in real terms). This could cause PVC B to be incorrectly deleted if it is actually still in use at (7).

The reproduction of this depends on the latency of APIServer failover, the potential risk of actual clock skew between APIServers, and the queue processing speed of the kube-controller-manager.

Potential enhancements:
1) Compare the admission time of a PVC to the PVCProtectionController workqueue (eg: the Y+N time) This would rely on the monotonically increasing clock of the `kube-controller-manager` process, rather than relying on the distributed clocks of the APIServer

**Comment by mattcary** [comment]
@pwschuurman I think you're right --- this is another reason to do the batching approach.

**Comment by pwschuurman** [comment]
/assign @pwschuurman 

> Open questions
>
>  * The controller uses a rate limiting workqueue. Do we have to change that to batch enough PVCs together? Probably we should write a PoC first and see how it works. The main point of rate limiting seems to be to not busy-retry failures, so as long as we have a rate limiter in the process work item frequency it should be fine.

I think the following would work in processWorkQueue:

1) Pull all PVC items off the workQueue into a list.
    * This ensures that all fetched PVCs are at least as old as the next subsequent live pod list call. The list needs to be scoped the same as the lazyLivePodList function definition.
2) In a loop, process each PVC item separately.
    * If any errors are encountered for an individual PVC, add it back to the workQueue in a rate limited manner. This retains the existing behavior of ensuring that a PVC that consistently produces an error when processed is queued through the rate limiter.

> call processPVC, adding a lazyLivePodList to the call

What happens if the first call to live Pod list fails? Should we continue to return an error for subsequent lazyLivePodList calls, or should we re-try the call right away? We may want to introduce a rate limiter within the function itself, to only retry the live() call after a certain amount of time has lapsed (this may already exist within the Kubernetes client definition). That would introduce two compounding rate limiters (one for PVCs that produce errors), one rate limiter for calls to the Kubernetes API server. There are some edge cases where we might delay more than expected depending on the number of PVCs in the workQueue, and then backoff frequency state of the rate limiter that aren't ideal. It would be nicer to have a separate backoff strategies for PVCs to be reprocessed, and for API call rate limiting and backoff.

**Comment by mattcary** [comment]
>What happens if the first call to live Pod list fails? Should we continue to return an error for subsequent lazyLivePodList 

Currently I think we requeue in a rate-limited fashion. Having some rate limiting within the lazy list seems to have the potential for complicated interactions as you point out.

Maybe we can just stall in the queue processing if a pod list error is signaled.

**Comment by serathius** [comment]
Looking into if https://github.com/kubernetes/enhancements/issues/2340 can remove the problem my making list all pods request much cheaper.

**Comment by mattcary** [comment]
I don't know if that will help --- it still goes out over the network, so will be called for every volume, which is too much.

**Comment by pwschuurman** [comment]
> Currently I think we requeue in a rate-limited fashion. Having some rate limiting within the lazy list seems to have the potential for complicated interactions as you point out.

To make the logic consistent (where failures in `askAPIServer` result in PVC being queued in a rate limiting fashion), we could continue to return the same error the same error in subsequent `lazyLivePodList` calls. This allows for a feedback loop of API Server list call faliures to result in rate limiting of PVC processing. If we don't do this (eg: we requeue without rate limiting), we may risk overloading the APIServer (we'd be dependent on client side throttling, based on the QPS that KCM has configured for the client, or compete for QPS with other clients in KCM.

I think this only causes a slight delay for PVCs to be reprocessed:
 * Previously: If we had N PVCs queued, and the 5th rate calls to list pods failed, only the 5th PVC would get requeued in a rate limited fashion. The 6th PVC would subsequently calls list pods, and it would all
 * New Behavior: If an error occurs on list pods for the 5th PVC, the 5th and 6th (up to N) PVCs will get requeued in a rate limited fashion. This introduces a small delay for the 6th PVC to get reprocessed (it needs to wait for rate limiting timeout), rather than getting reprocessed right away in the next loop iteration. If this is the first time getting requeued, this should be a short delay ([5ms, which is the default initial delay](https://github.com/kubernetes/client-go/blob/d3682da14f7208b5bb30bc2ae94d44606a77abac/util/workqueue/default_rate_limiters.go#L52)). So I think this should be acceptable.

We could potentially requeue with rate limiting based on the error rate (eg: if unavailable, or timeout, requeue right away), but I'm concerned this could lead to a situation where the client misbehaves and sends a large number of API requests against the APIServer, up to the QPS limit.

**Comment by mattcary** [comment]
Yeah that's a good point. We definitely want to rate limit live list calls.

I think we can work through the mechanics on the PR --- the goal should remain though to be able to delete 10s of k PVCs in minutes.
